spec/controllers/v0/user_action_events_controller_spec.rb

```diff
@@ -0,0 +1,206 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe V0::UserActionEventsController, type: :controller do
+  include RequestHelper
+
+  describe 'GET #index' do
+    subject { get(:index, params: index_params) }
+
+    let(:index_params) { { start_date:, end_date:, page:, per_page: } }
+    let(:start_date) { nil }
+    let(:end_date) { nil }
+    let(:page) { nil }
+    let(:per_page) { nil }
+
+    context 'when not logged in' do
+      it 'returns unauthorized' do
+        get :index
+        expect(response).to have_http_status(:unauthorized)
+      end
+    end
+
+    context 'when logged in' do
+      let(:idme_uuid) { 'some-idme-uuid' }
+      let(:user) { create(:user, idme_uuid:) }
+      let!(:user_verification) { create(:idme_user_verification, idme_uuid:) }
+
+      before do
+        sign_in_as(user)
+      end
+
+      context 'when there are user actions' do
+        let(:user_action_event_one) { create(:user_action_event, details: 'Sample event 1') }
+        let(:user_action_event_two) { create(:user_action_event, details: 'Sample event 2') }
+
+        let!(:user_action_one) do
+          create(:user_action, subject_user_verification_id: user_verification.id,
+                               user_action_event: user_action_event_one, created_at: 1.year.ago)
+        end
+        let!(:user_action_two) do
+          create(:user_action, subject_user_verification_id: user_verification.id,
+                               user_action_event: user_action_event_two, created_at: 2.months.ago)
+        end
+        let!(:user_action_three) do
+          create(:user_action, subject_user_verification_id: user_verification.id,
+                               user_action_event: user_action_event_one, created_at: 3.weeks.ago)
+        end
+        let!(:user_action_four) do
+          create(:user_action, subject_user_verification_id: user_verification.id,
+                               user_action_event: user_action_event_two, created_at: 4.days.ago)
+        end
+
+        context 'when filtering by date range' do
+          let(:start_date) { 3.months.ago.to_date }
+          let(:end_date) { 2.weeks.ago.to_date }
+
+          it 'returns the results in descending order by created_at' do
+            subject
```

Comment:
You only need to call `subject` once, so this test (and all of the other ones that have a line like this) can just begin with `json_response = JSON.parse(subject.body)` and then you can set your expectations against that `json_response` variable.

If you call the subject like that and set it to a variable then you won't need `subject` or `get :index` declared at the top of any of your tests.